fe/fe-core/src/main/java/org/apache/doris/statistics/OlapAnalysisTask.java

```diff
@@ -194,22 +226,66 @@ protected Pair<List<Long>, Long> calcActualSampleTablets() {
                 long tabletId = ids.get(seekTid);
                 sampleTabletIds.add(tabletId);
                 actualSampledRowCount += baseIndex.getTablet(tabletId).getRowCount(true);
+                if (actualSampledRowCount >= sampleRows && !forPartitionColumn) {
+                    enough = true;
+                    break;
+                }
             }
-
             totalRows += p.getBaseIndex().getRowCount();
             totalTablet += ids.size();
+            if (enough) {
+                break;
+            }
         }
 
         // all hit, direct full
         if (totalRows < sampleRows) {
             // can't fill full sample rows
             sampleTabletIds.clear();
-        } else if (sampleTabletIds.size() == totalTablet) {
-            // TODO add limit
+        } else if (sampleTabletIds.size() == totalTablet && !enough) {
             sampleTabletIds.clear();
-        } else if (!sampleTabletIds.isEmpty()) {
-            // TODO add limit
         }
         return Pair.of(sampleTabletIds, actualSampledRowCount);
     }
+
+    /**
+     * For ordinary column (neither key column nor partition column), need to limit sample size to user specified value.
+     * @return Return true when need to limit.
+     */
+    protected boolean needLimit() {
+        // Key column is sorted, use limit will cause the ndv not accurate enough, so skip key columns.
+        if (col.isKey()) {
+            return false;
+        }
+        // Partition column need to scan tablets from all partitions.
+        if (isPartitionColumn()) {
+            return false;
+        }
+        return true;
+    }
+
+    protected boolean isDistributionColumn() {
```

Comment:
I think we can add 2 new interface: `isDistributionColumn(String colName)` and `isPartitionColumn(String colName)` in `TableIf`?
So that when implement other type of AnalysisTask, we can just use that interface.